dynamic/longestcommonsubsequence.go

```diff
@@ -8,10 +8,14 @@ import (
 	"unicode/utf8"
 )
 
+func extractData(inStr string) (int, []rune) {
```

Comment:
This function seems unnecessary and obfuscates the code. The rune count should equal the length of the rune array.

Reply:
> The rune count should equal the length of the rune array.

Please have a look [here](https://go.dev/play/p/JKbfU7z0o4o).

The problem is that if you have `someString` (of the type `string`), then with `someString[i]` you access the byte at the position `i`.

Reply:
I understand that, however, you are converting it to a `[]rune` so it should not be a problem.

Reply:
Exactly. That is the reason to have the function `extractData` - it is not unnecessary.

Reply:
You can just do:
```go
A := []rune(a)
aLen := len(A)
```
I don't see how that function is useful or more intuitive.

Reply:
The same has to be done for `b`. The function `extractData` helps to avoid code duplication.

Reply:
This duplication is not so severe that we must sacrifice readability and intuitiveness for it.

Reply:
I agree with @raklaptudirm

Reply:
Having even a _trivial_ duplication like this 
```golang
A := []rune(a)
aLen := len(A)
B := []rune(b)
bLen := len(B)
```
is a bad idea because of (at least) two reasons:
- possibility of having a typo,
- these lines are not _testable_. One can write tests for the function `extractData`, but it is not possible to test the lines above separately.

In the discussed case, we really want to do the same thing for the strings `a` and `b`: we want to compute their length and cast it to `[]rune`. It is not just a coincidence, that the _code looks the same_. We want to do the exact same thing. That motivates the introduction of this function - having it underlines that _we really do the same thing_. Sometimes it happens that the code _looks the same_ but the purpose of it is slightly different - abstracting it into a single function would be _bad_.

@raklaptudirm I have to admit that I like your idea of getting rid of `utf8.RuneCountInString` while  computing the number of _characters_.

After thinking about it, I think that the best way would be to change `LongestCommonSubsequence` to accept `[]rune`, i.e. to have `LongestCommonSubsequence([]rune a, []rune b) int`. Then it is clear just from looking at the signature that this function only _solves the problem_ and how does it handle the _multibyte_ characters. @tjgurwara99 @raklaptudirm please let me know what do you think about that.

Reply:
The problem I have with your function `extractData` is that it's not really a great name for a function and the order of `return` types. I argue that it's really not conveying what it does when reading the code. Questions like "what does it extract?" or "why does it extract?" pop up as soon as I read it which means before I can read further I have to step into the code and read that code causing me to switch context a few more times than I need to. The simplest approach is quite straight forward imo as @raklaptudirm mentioned - but if you want to use a function, calling it something self explanatory is better than calling it `extractData`. Something like `strToRuneSlice(s string) (r []rune, size int)` is much clearer to me at least.

> these lines are not testable. One can write tests for the function `extractData`, but it is not possible to test the lines above separately.

Why would you want to test something like that separately?

> possibility of having a typo,

That possibility exists everywhere - why do you think PR reviews exist? 😂 I can do this

```Go
extractData(a)
extratcData(b)
```

The compiler would complain in the above case and any other type that you may want to think of in this context.

Reply:
Go's type system almost completely removes the possibility of typos going undetected, because either the symbol doesn't exist or it's type is different. One of the two cases is true a majority of the time.

Reply:
@tjgurwara99 thank you for the idea for better name, I really like the `strToRuneSlice`.

> That possibility exists everywhere - why do you think PR reviews exist?

The code review is about the things, which cannot be verified by automated tools. I was more thinking about the typo like:
```golang
A := []rune(b)
aLen := len(A)
B := []rune(a)
bLen := len(B)
```
Above would not be detected by the compiler or other tools - all of the variables are used etc. but of course the _logic_ behind the names is broken. Note that the result of `LongestCommonSubsequence` will remain correct but the potential reader of the code would be a bit confused.


> Why would you want to test something like that separately?

In this _simple case_, I would not. But is is a good practice to write _testable_ code.


@raklaptudirm @tjgurwara99 I think I can mark this as resolved.

Reply:
I still don't think sacrificing readability is a good idea. Typos can happen everywhere as previously stated.

Reply:
I do not understand why introducing this function reduces the readability. It is like:
```golang
aRunes, aLen := strToRuneSlice(a)
bRunes, bLen := strToRuneSlice(b)
```
vs.
```golang
aRunes := []rune(a)
aLen := len(A)
bRunes := []rune(b)
bLen := len(B)
```
I think that a person, who does know the `Go` syntax is more likely to have less problems with reading the first one. In the first case there is this extra bit of information that _we really intend to do the exact same thing to `a` and `b`_. Moreover the first one is simply twice shorter.

Reply:
I think if you use the `strToRuneSlice`, I don't mind personally.

> I think that a person, who does know the Go syntax is more likely to have less problems with reading the first one.

That is under the assumption that this is being **read** by a person who does indeed know Go, and its syntax and terminology (eg `Rune` and `Slice`). But this repository is for learning purposes, so you can see why that `strToRuneSlice` would require the person not familiar with Go to step into the code and see (albeit its in the same file so I personally don't mind) but still... 😄

Reply:
@raklaptudirm Can I mark this conversation as resolved?